### PR TITLE
introduce TokenStream interface

### LINKED ISSUE #1094: Extract token stream interface from parser
### Issue To Be Solved

The parser-guided fuzzer needs to have a couple alternate implementations of token buffering and token access, and it needs to be switchable at time of Parse().  Previously I tried to use a buildtag to isolate the different implementations, but the parser-guided fuzzer needs to switch at runtime, and needs more than 2 implementations which makes the buildtag approach even more difficult.

### Tradeoff

* There could be a performance hit due to the indirection, but existing prototype work shows no noticeable change in the running time for the full cadence test suite.

### Context

The parser-guided-fuzzer is blocked on this.

### PR DESCRIPTION
**Pull request by jsatdapr**
Closes #1094

## Description

We want to write fuzzers with custom TokenStream implementations.  This PR introduces a TokenStream interface and a default implementation.
______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by SupunS — COMMENTED** [review]
Looks good! just a couple of questions/comments.

File: runtime/parser2/lexer/lexer.go [hunk 1]
```diff
@@ -74,9 +76,44 @@ type lexer struct {
...
+			Range: ast.Range{
+				StartPos: pos,
+				EndPos:   pos,
+			},
+		}
+	}
+	return token
+}
+
+func (l *lexer) Close(_ *[]error) {
```
[the remark this answers is not in the document]
Comment (jsatdapr) [reply runtime/parser2/lexer/lexer.go:104]: We don't need errors now, true.  I will propose it again on another PR when it becomes necessary.

File: runtime/parser2/lexer/lexer.go [hunk 2]
```diff
@@ -74,9 +76,44 @@ type lexer struct {
 	startPos position
 }
 
-func Lex(ctx context.Context, input string) chan Token {
+type TokenStream interface {
+	Next() Token
```
[the remark this answers is not in the document]
Comment (jsatdapr) [reply runtime/parser2/lexer/lexer.go:80]: Added documentation for Next and others.

File: runtime/parser2/lexer/lexer.go [hunk 3]
```diff
@@ -74,9 +76,47 @@ type lexer struct {
...
+	Code() string
+}
+
+func (l *lexer) Next() Token {
+	token, ok := <-l.tokens
+	if !ok {
+		pos := ast.Position{
+			Offset: l.endOffset - 1,
+			Line:   l.endPos().line,
+			Column: l.endPos().column - 1,
```
Comment (turbolent) [inline runtime/parser2/lexer/lexer.go:94]: Maybe refactor `l.endPos()` into a variable, so it is only computed once

File: runtime/parser2/lexer/lexer.go [hunk 4]
```diff
@@ -74,9 +76,47 @@ type lexer struct {
...
+	}
+	return token
+}
+
+func (l *lexer) Close() {
+	l.cancelLexer()
+}
+
+func (l *lexer) Code() string {
+	return l.input
```
Comment (turbolent) [inline runtime/parser2/lexer/lexer.go:112]: From the documentation of `Code` it seemed to me at first that the function would only return the lexed code up-to when `Close` was called

**Review by turbolent — APPROVED** [review]
Nice!

File: runtime/parser2/lexer/lexer.go (hunk 3 above)
Comment (turbolent) [reply runtime/parser2/lexer/lexer.go]: Refactored in fd6b5b3

File: runtime/parser2/lexer/lexer.go (hunk 4 above)
Comment (turbolent) [reply runtime/parser2/lexer/lexer.go:113]: 313b77b

**Pull request merged**